src/test/scala/org/apache/spark/sql/delta/MergeIntoSuiteBase.scala

```diff
@@ -1724,6 +1748,118 @@ abstract class MergeIntoSuiteBase
     }
   }
 
+  def createTableHelper(
+      df: DataFrame,
+      partitionColumns: Seq[String],
+      basePath: String): Unit = {
+    df.write
+      .format("delta")
+      .partitionBy(partitionColumns: _*)
+      .save(basePath)
+  }
+
+  test("auto repartition merge - partition on multiple columns") {
+    withTempDir { tempDir =>
+      // Create two copies of the target table
+      val tgt1 = tempDir.getAbsolutePath + "target"
+      val tgt2 = tempDir.getAbsolutePath + "targetRepartitioned"
+      val df = spark.range(100)
+        .withColumn("part", 'id % 5)
+        .withColumn("updated", 'id % 3)
+      createTableHelper(df, Seq("part", "updated"), tgt1)
+      createTableHelper(df, Seq("part", "updated"), tgt2)
+
+      // Create source as a view
+      spark.range(80, 110)
+        .withColumn("part", 'id % 5)
+        .withColumn("updated", 'id % 3)
+        .createOrReplaceTempView("source")
+
+      val deltaLogTgt1 = DeltaLog.forTable(spark, tgt1)
+      val deltaLogTgt2 = DeltaLog.forTable(spark, tgt2)
+      val filesBeforeMerge = deltaLogTgt1.snapshot.numOfFiles
+
+      executeMergesForRepartitionTests(
+        s"delta.`$tgt1` as t",
+        s"delta.`$tgt2` as t",
+        "source src",
+        cond = "src.id = t.id",
+        update("t.updated = 1"),
+        insert("(id, part, updated) VALUES (id, part, updated)")
+      )
+
+      val filesAfterNoRepartition = deltaLogTgt1.snapshot.numOfFiles
+      val filesAfterRepartition = deltaLogTgt2.snapshot.numOfFiles
+
+      // Check if the number of files in the target with repartition enabled is
+      // lesser than the other target
+      assert(filesAfterNoRepartition > filesAfterRepartition)
+
+      // The number of files after the merge is same as the number of files before as all the
+      // partitions are updated hence the inserts and the updates are rewritten as new files.
+      assert(filesBeforeMerge == filesAfterRepartition)
```

Comment:
I feel like these additional asserts outside `runMergeWithRepartitionTest` are not useful in testing what we want to test. What we want to test is whether repartition reduces the number of files written or not. It does not need to test how many extra files it wrote compared to before, etc. Those are specific behavior of different types of merge and relevant here. Whether in all it replace 5 files or added 5 new files .. does not matter as long as repartitioning reduces the number new files written.